server/src/main/java/org/elasticsearch/search/query/QueryPhaseCollector.java

```diff
@@ -0,0 +1,336 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.search.query;
+
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.search.CollectionTerminatedException;
+import org.apache.lucene.search.Collector;
+import org.apache.lucene.search.DocIdSetIterator;
+import org.apache.lucene.search.FilterScorable;
+import org.apache.lucene.search.LeafCollector;
+import org.apache.lucene.search.Scorable;
+import org.apache.lucene.search.ScoreCachingWrappingScorer;
+import org.apache.lucene.search.ScoreMode;
+import org.apache.lucene.search.ScorerSupplier;
+import org.apache.lucene.search.Weight;
+import org.apache.lucene.util.Bits;
+import org.elasticsearch.common.lucene.Lucene;
+
+import java.io.IOException;
+import java.util.Objects;
+
+/**
+ * Top-level collector used in the query phase to perform top hits collection as well as aggs collection.
+ * Inspired by {@link org.apache.lucene.search.MultiCollector} but specialized for wrapping two collectors and filtering collected
+ * documents as follows:
+ * - through an optional <code>post_filter</code> that is applied to the top hits collection
+ * - through an optional <code>min_score</code> threshold, which is applied to both the top hits as well as aggs.
+ * Supports also terminating the collection after a certain number of documents have been collected (<code>terminate_after</code>).
+ *
+ * When top docs as well as aggs are collected (because both collectors were provided), skipping low scoring hits via
+ * {@link Scorable#setMinCompetitiveScore(float)} is not supported for either of the collectors.
+ */
+final class QueryPhaseCollector implements Collector {
+    private final Collector aggsCollector;
+    private final Collector topDocsCollector;
+    private final int terminateAfter;
+    private final Weight postFilterWeight;
+    private final Float minScore;
+    private final boolean cacheScores;
+
+    private int numCollected;
+    private boolean terminatedAfter = false;
+
+    QueryPhaseCollector(Collector topDocsCollector, Weight postFilterWeight, int terminateAfter, Collector aggsCollector, Float minScore) {
+        this.topDocsCollector = Objects.requireNonNull(topDocsCollector);
```

Comment:
I am wondering if the aggsCollector should be a BucketCollector as it does not need anymore to implement the Collector interface.

Reply:
I would argue that even the TopDocsCollector shouldn't be a Collector either. Maybe there is a chance to find a better abstraction for Elasticsearch Collectors that it is not tangled with Lucene abstraction.